File: web/app/src/acheron/passwordGenerator.js
/**
 * Generador de contraseñas aleatorias, 100% local (Web Crypto API).
 *
 * No se comunica con el servidor: coherente con el resto de Acheron, donde
 * el cifrado y las claves nunca salen del navegador. Usa la misma fuente de
 * aleatoriedad segura que `crypto.js` (`crypto.getRandomValues`), con
 * rechazo uniforme para no sesgar el módulo hacia ningún carácter.
 */

import { randomBytes } from './crypto.js'

const UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
const LOWER = 'abcdefghijklmnopqrstuvwxyz'
const DIGITS = '0123456789'
const SYMBOLS = '!@#$%^&*()-_=+[]{};:,.<>?'
const AMBIGUOUS = new Set('0O1lI')

/** Entero uniforme en [0, max) usando bytes criptográficos (rechazo de sesgo por módulo). */
function secureInt(max) {
  const limit = 256 - (256 % max)
  let byte
  do {
    byte = randomBytes(1)[0]
  } while (byte >= limit)
  return byte % max
}

function secureChoice(str) {
  return str[secureInt(str.length)]
}

function secureShuffle(arr) {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = secureInt(i + 1)
    ;[arr[i], arr[j]] = [arr[j], arr[i]]
  }
  return arr
}

/**
 * Genera una contraseña aleatoria segura.
 *
 * @param {object} [options]
 * @param {number} [options.length=20]
 * @param {boolean} [options.uppercase=true]
 * @param {boolean} [options.lowercase=true]
 * @param {boolean} [options.digits=true]
 * @param {boolean} [options.symbols=true]
 * @param {boolean} [options.excludeAmbiguous=true] excluye 0/O, 1/l/I
 * @returns {string}
 */
export function generatePassword({
  length = 20,
  uppercase = true,
  lowercase = true,
  digits = true,
  symbols = true,
  excludeAmbiguous = true,
} = {}) {
  let pools = [uppercase && UPPER, lowercase && LOWER, digits && DIGITS, symbols && SYMBOLS].filter(
    Boolean,
  )
  if (excludeAmbiguous) {
    pools = pools.map((pool) => [...pool].filter((c) => !AMBIGUOUS.has(c)).join(''))
  }

  const alphabet = pools.join('')
  const chars = pools.map((pool) => secureChoice(pool)) // >=1 carácter de cada set activado
  while (chars.length < length) chars.push(secureChoice(alphabet))

  return secureShuffle(chars).join('')
}

File: API/src/modules/acheron/password_generator.py
"""Generador de contraseñas aleatorias seguras.

Utilidad sin estado: no depende de vaults, usuarios ni base de datos.
Usa `secrets` (CSPRNG) para toda la aleatoriedad, igual que el resto
de generación de secretos en la API (ver users/managers.py).
"""

import secrets
import string

_AMBIGUOUS = set("0O1lI")
_SYMBOLS = "!@#$%^&*()-_=+[]{};:,.<>?"


def generate_password(
    length: int,
    uppercase: bool,
    lowercase: bool,
    digits: bool,
    symbols: bool,
    exclude_ambiguous: bool,
) -> str:
    pools = []
    if uppercase:
        pools.append(string.ascii_uppercase)
    if lowercase:
        pools.append(string.ascii_lowercase)
    if digits:
        pools.append(string.digits)
    if symbols:
        pools.append(_SYMBOLS)

    if exclude_ambiguous:
        pools = ["".join(c for c in pool if c not in _AMBIGUOUS) for pool in pools]

    alphabet = "".join(pools)
    chars = [secrets.choice(pool) for pool in pools]
    chars += [secrets.choice(alphabet) for _ in range(length - len(chars))]
    secrets.SystemRandom().shuffle(chars)
    return "".join(chars)
